# Objective

The aim is to leverage SQL querying and data manipulation techniques to extract meaningful information, identify patterns, and generate actionable recommendations to support evidence-based decision-making in various domains, such as public policy, and resource allocation.

## Importing Library 

In [2]:
import pandas as pd
import sqlite3


In [3]:
!pip install ipython-sql


## Loading dataset 

In [4]:
df_1=pd.read_excel('Dataset1.xlsx')
df_2=pd.read_excel('Dataset2.xlsx')


In [5]:
df_1.head()

District        State  Growth  Sex_Ratio  Literacy
0                       Thane  Maharashtra  0.3601        886     84.53
1  North Twenty Four Parganas  West Bengal  0.1204        955     84.06
2                   Bangalore    Karnataka  0.4718        916     87.67
3                        Pune  Maharashtra  0.3037        915     86.15
4             Mumbai Suburban  Maharashtra  0.0829        860     89.91

In [6]:
df_2.head()

District           State  Area_km2  Population
0    Adilabad  Andhra Pradesh     16105     2741239
1        Agra   Uttar Pradesh      4041     4418797
2   Ahmadabad         Gujarat      8107     7214225
3  Ahmadnagar     Maharashtra     17048     4543159
4      Aizawl         Mizoram      3576      400309

In [7]:
connection=sqlite3.connect('Indian_census_dataset.db')

In [8]:
df_1.to_sql('Dataset_1',connection)


640

In [9]:
df_2.to_sql('Dataset_2',connection)

640

In [10]:
%reload_ext sql


In [11]:
%sql sqlite:///Indian_census_dataset.db

In [12]:
 %%sql
select * from Dataset_1 limit 10

 * sqlite:///Indian_census_dataset.db
Done.


[(0, 'Thane', 'Maharashtra', 0.3601, 886, 84.53),
 (1, 'North Twenty Four Parganas', 'West Bengal', 0.1204, 955, 84.06),
 (2, 'Bangalore', 'Karnataka', 0.4718, 916, 87.67),
 (3, 'Pune', 'Maharashtra', 0.3037, 915, 86.15),
 (4, 'Mumbai Suburban', 'Maharashtra', 0.0829, 860, 89.91),
 (5, 'South Twenty Four Parganas', 'West Bengal', 0.1817, 956, 77.51),
 (6, 'Barddhaman', 'West Bengal', 0.1192, 945, 76.21),
 (7, 'Ahmadabad', 'Gujarat', 0.2403, 904, 85.31),
 (8, 'Murshidabad', 'West Bengal', 0.2109, 958, 66.59),
 (9, 'Jaipur', 'Rajasthan', 0.2619, 910, 75.51)]

In [13]:
%%sql
select * from Dataset_2 limit 10

 * sqlite:///Indian_census_dataset.db
Done.


[(0, 'Adilabad', 'Andhra Pradesh', 16105, 2741239),
 (1, 'Agra', 'Uttar Pradesh', 4041, 4418797),
 (2, 'Ahmadabad', 'Gujarat', 8107, 7214225),
 (3, 'Ahmadnagar', 'Maharashtra', 17048, 4543159),
 (4, 'Aizawl', 'Mizoram', 3576, 400309),
 (5, 'Ajmer', 'Rajasthan', 8481, 2583052),
 (6, 'Akola', 'Maharashtra', 5676, 1813906),
 (7, 'Alappuzha', 'Kerala', 1415, 2127789),
 (8, 'Aligarh', 'Uttar Pradesh', 3650, 3673889),
 (9, 'Alirajpur', 'Madhya Pradesh', 3182, 728999)]

## 01. Number of rows in dataset1


In [14]:
%%sql
select count(*)as total_rows from Dataset_1

 * sqlite:///Indian_census_dataset.db
Done.


[(640,)]

## 02. Number of rows in dataset2


In [15]:
%%sql
select count(*)as total_rows from Dataset_2

 * sqlite:///Indian_census_dataset.db
Done.


[(640,)]

## 03. Dataset for bihar and jharkhand


In [16]:
%%sql
select * from Dataset_1 where State in ('Bihar' , 'Jharkhand') limit 10

 * sqlite:///Indian_census_dataset.db
Done.


[(14, 'Patna', 'Bihar', 0.2373, 897, 70.68),
 (19, 'Purbi Champaran', 'Bihar', 0.2943, 902, 55.79),
 (23, 'Muzaffarpur', 'Bihar', 0.2814, 900, 63.43),
 (36, 'Madhubani', 'Bihar', 0.2551, 926, 58.62),
 (41, 'Gaya', 'Bihar', 0.2643, 937, 63.67),
 (44, 'Samastipur', 'Bihar', 0.2553, 911, 61.86),
 (58, 'Saran', 'Bihar', 0.2164, 954, 65.96),
 (60, 'Darbhanga', 'Bihar', 0.1947, 911, 56.56),
 (63, 'Pashchim Champaran', 'Bihar', 0.2929, 909, 55.7),
 (86, 'Vaishali', 'Bihar', 0.2857, 895, 66.6)]

## 03. Population of India

In [17]:
%%sql
select sum(Population) as Population from Dataset_2


 * sqlite:///Indian_census_dataset.db
Done.


[(1210854977,)]

##  04. Average growth of india

In [18]:
%%sql
select avg(growth)*100 as Average_growth from Dataset_1

 * sqlite:///Indian_census_dataset.db
Done.


[(19.24592187500001,)]

## 05. Average growth of different states
 

In [19]:
%%sql
select  State  , avg(growth)*100 as Average_growth from Dataset_1 group by State order by Average_growth desc 


 * sqlite:///Indian_census_dataset.db
Done.


[('Nagaland', 82.28090909090909),
 ('Dadra and Nagar Haveli', 55.879999999999995),
 ('Daman and Diu', 42.74),
 ('Puducherry', 34.2975),
 ('Meghalaya', 30.39428571428572),
 ('Arunachal Pradesh', 27.805624999999996),
 ('Manipur', 25.54111111111111),
 ('Bihar', 25.23184210526316),
 ('Mizoram', 24.346249999999998),
 ('Jammu and Kashmir', 24.270454545454545),
 ('Jharkhand', 23.78166666666666),
 ('Rajasthan', 21.125454545454545),
 ('Haryana', 20.492380952380948),
 ('Madhya Pradesh', 20.242199999999993),
 ('Uttar Pradesh', 20.037605633802823),
 ('Chhattisgarh', 20.035),
 ('Chandigarh', 17.19),
 ('Gujarat', 17.082307692307694),
 ('Tripura', 16.7525),
 ('Assam', 16.523703703703703),
 ('Tamil Nadu', 14.6215625),
 ('Orissa', 14.24933333333334),
 ('West Bengal', 14.132105263157897),
 ('Maharashtra', 13.945428571428572),
 ('Delhi', 13.887777777777776),
 ('Punjab', 13.680499999999999),
 ('Uttarakhand', 13.057692307692308),
 ('Karnataka', 11.859333333333332),
 ('Himachal Pradesh', 11.3425),
 ('Sikkim', 11.149999999999999),
 ('Andhra Pradesh', 10.923913043478262),
 ('Goa', 8.285),
 ('Lakshadweep', 6.3),
 ('Kerala', 4.132142857142857),
 ('Andaman And Nicobar Islands', 0.5966666666666666)]

## 06. Average sex ratio of different states 


In [20]:
%%sql
select State , round(avg(Sex_Ratio),0) as avg_sex_ratio from Dataset_1 group by 
State  order by avg_sex_ratio desc

 * sqlite:///Indian_census_dataset.db
Done.


[('Kerala', 1080.0),
 ('Puducherry', 1075.0),
 ('Uttarakhand', 1010.0),
 ('Tamil Nadu', 999.0),
 ('Chhattisgarh', 995.0),
 ('Andhra Pradesh', 995.0),
 ('Orissa', 984.0),
 ('Karnataka', 984.0),
 ('Meghalaya', 980.0),
 ('Manipur', 976.0),
 ('Goa', 975.0),
 ('Mizoram', 966.0),
 ('Tripura', 958.0),
 ('Assam', 958.0),
 ('Jharkhand', 956.0),
 ('Himachal Pradesh', 953.0),
 ('West Bengal', 950.0),
 ('Maharashtra', 948.0),
 ('Lakshadweep', 946.0),
 ('Gujarat', 938.0),
 ('Madhya Pradesh', 936.0),
 ('Nagaland', 934.0),
 ('Rajasthan', 930.0),
 ('Arunachal Pradesh', 920.0),
 ('Bihar', 918.0),
 ('Uttar Pradesh', 909.0),
 ('Punjab', 897.0),
 ('Jammu and Kashmir', 883.0),
 ('Haryana', 880.0),
 ('Sikkim', 874.0),
 ('Delhi', 866.0),
 ('Andaman And Nicobar Islands', 858.0),
 ('Chandigarh', 818.0),
 ('Daman and Diu', 783.0),
 ('Dadra and Nagar Haveli', 774.0)]

## 07. The states which average Literacy ratio is geater than 90

In [21]:
%%sql

select State , round(avg(Literacy),0) as avg_literacy_ratio from Dataset_1 group by 
State  having  avg_literacy_ratio >90 order by avg_literacy_ratio desc

 * sqlite:///Indian_census_dataset.db
Done.


[('Kerala', 94.0), ('Lakshadweep', 92.0)]

## 08. Top three states with highest growth

In [22]:
%%sql
select   State  , avg(growth)*100 as Average_growth from Dataset_1 group by 
State  order by Average_growth desc limit 3




 * sqlite:///Indian_census_dataset.db
Done.


[('Nagaland', 82.28090909090909),
 ('Dadra and Nagar Haveli', 55.879999999999995),
 ('Daman and Diu', 42.74)]

## 11. States which starting with 'a' and ending with 'm'
 

In [23]:
%%sql

select distinct State from Dataset_1 where lower(state) like 'a%' and lower (state) 
like '%m'

 * sqlite:///Indian_census_dataset.db
Done.


[('Assam',)]

## 12. Joining both table 

In [24]:
%%sql
select a.District , a.State , a.Sex_Ratio , b.Population from Dataset_1 as a inner 
join Dataset_2 as b on a.District = b.District limit 10


 * sqlite:///Indian_census_dataset.db
Done.


[('Thane', 'Maharashtra', 886, 11060148),
 ('North Twenty Four Parganas', 'West Bengal', 955, 10009781),
 ('Bangalore', 'Karnataka', 916, 9621551),
 ('Pune', 'Maharashtra', 915, 9429408),
 ('Mumbai Suburban', 'Maharashtra', 860, 9356962),
 ('South Twenty Four Parganas', 'West Bengal', 956, 8161961),
 ('Barddhaman', 'West Bengal', 945, 7717563),
 ('Ahmadabad', 'Gujarat', 904, 7214225),
 ('Murshidabad', 'West Bengal', 958, 7103807),
 ('Jaipur', 'Rajasthan', 910, 6626178)]

## 13. Finding number of males and females in different states

In [25]:
%%sql

select d.state , sum(d.males)as total_males, sum(d.females) as total_females from

(select c.district , c.state , round((c.population)/(c.sex_ratio+1),0) as males , 
round((c.population*c.sex_ratio)/(c.sex_ratio+1) ,0) as females from
 
(select a.District , a.State , a.Sex_Ratio/1000 as sex_ratio, b.Population from 
Dataset_1 as a inner join Dataset_2 as b on a.District = b.District) as c)

as d

group by d.state ;


 * sqlite:///Indian_census_dataset.db
Done.


[('Andaman And Nicobar Islands', 274984.0, 0.0),
 ('Andhra Pradesh', 66989042.0, 17591730.0),
 ('Arunachal Pradesh', 1121771.0, 85383.0),
 ('Assam', 31205576.0, 0.0),
 ('Bihar', 101420357.0, 1281006.0),
 ('Chandigarh', 1055450.0, 0.0),
 ('Chhattisgarh', 24360212.0, 2817266.0),
 ('Daman and Diu', 217210.0, 26037.0),
 ('Delhi', 142004.0, 0.0),
 ('Goa', 1458545.0, 0.0),
 ('Gujarat', 55404173.0, 517656.0),
 ('Haryana', 25351462.0, 0.0),
 ('Himachal Pradesh', 8566498.0, 2034451.0),
 ('Jammu and Kashmir', 12541302.0, 0.0),
 ('Jharkhand', 31171909.0, 751169.0),
 ('Karnataka', 57018163.0, 4332359.0),
 ('Kerala', 16703027.0, 16703027.0),
 ('Lakshadweep', 64473.0, 0.0),
 ('Madhya Pradesh', 65586263.0, 2095062.0),
 ('Maharashtra', 112090618.0, 1232359.0),
 ('Manipur', 2157657.0, 698136.0),
 ('Meghalaya', 2097526.0, 610523.0),
 ('Mizoram', 897051.0, 200154.0),
 ('Nagaland', 1978502.0, 0.0),
 ('Orissa', 36154229.0, 5819984.0),
 ('Puducherry', 623976.0, 623976.0),
 ('Punjab', 26748710.0, 0.0),
 ('Rajasthan', 71757578.0, 0.0),
 ('Tamil Nadu', 57086624.0, 15060398.0),
 ('Tripura', 3673917.0, 0.0),
 ('Uttar Pradesh', 190674944.0, 6104531.0),
 ('Uttarakhand', 6199265.0, 1309330.0),
 ('West Bengal', 91276115.0, 0.0)]

## 14. Total literacy ratio of different states

In [26]:
%%sql

select c.state,sum(c.literate_people) as total_literate_people,sum
(c.illiterate_people) as total_illiterate_people from

(select d.district,d.state,round( d.literacy_ratio*d.population,0) as 
literate_people,round((1-d.literacy_ratio)*d.population,0) as illiterate_people 
from 
 
(select a.district,a.state,a.literacy/100 as literacy_ratio ,b.population from 
Dataset_1 as a inner join Dataset_2 as b on a.District=b.District ) as d)

as c

group by c.state

 * sqlite:///Indian_census_dataset.db
Done.


[('Andaman And Nicobar Islands', 241015.0, 33969.0),
 ('Andhra Pradesh', 56671677.0, 27909100.0),
 ('Arunachal Pradesh', 761557.0, 445597.0),
 ('Assam', 22484409.0, 8721167.0),
 ('Bihar', 63994271.0, 38707092.0),
 ('Chandigarh', 908215.0, 147235.0),
 ('Chhattisgarh', 18621154.0, 8556328.0),
 ('Daman and Diu', 211827.0, 31420.0),
 ('Delhi', 125446.0, 16558.0),
 ('Goa', 1293736.0, 164809.0),
 ('Gujarat', 44139127.0, 11782703.0),
 ('Haryana', 19123489.0, 6227973.0),
 ('Himachal Pradesh', 8883924.0, 1717028.0),
 ('Jammu and Kashmir', 8391150.0, 4150152.0),
 ('Jharkhand', 21132066.0, 10791012.0),
 ('Karnataka', 46146322.0, 15204205.0),
 ('Kerala', 31395554.0, 2010507.0),
 ('Lakshadweep', 59218.0, 5255.0),
 ('Madhya Pradesh', 47069447.0, 20611880.0),
 ('Maharashtra', 92954049.0, 20368930.0),
 ('Manipur', 2196539.0, 659255.0),
 ('Meghalaya', 2009086.0, 698963.0),
 ('Mizoram', 999927.0, 97279.0),
 ('Nagaland', 1569750.0, 408752.0),
 ('Orissa', 30486803.0, 11487415.0),
 ('Puducherry', 1071351.0, 176602.0),
 ('Punjab', 20201266.0, 6547444.0),
 ('Rajasthan', 47063500.0, 24694078.0),
 ('Tamil Nadu', 57775522.0, 14371508.0),
 ('Tripura', 3203603.0, 470314.0),
 ('Uttar Pradesh', 133084739.0, 63694737.0),
 ('Uttarakhand', 5992810.0, 1515789.0),
 ('West Bengal', 69475388.0, 21800727.0)]

## 15. Population in previous Census

In [27]:
%%sql

select d.state,sum(d.prev_populatoin) as prev_census_population , sum
(d.curr_population) as curr_census_population from

(select c.district ,c.state, round(( c.population- c.growth*c.population ),0) as 
prev_populatoin , c.growth, c.population as curr_population from 
 
(select a.district,a.state,a.Growth as Growth ,b.population from Dataset_1 as a inner 
join Dataset_2 as b on a.District=b.District)as c) 

as d 

group by d.state


 * sqlite:///Indian_census_dataset.db
Done.


[('Andaman And Nicobar Islands', 245672.0, 274984),
 ('Andhra Pradesh', 74686197.0, 84580777),
 ('Arunachal Pradesh', 866114.0, 1207154),
 ('Assam', 25807225.0, 31205576),
 ('Bihar', 76674920.0, 102701363),
 ('Chandigarh', 874018.0, 1055450),
 ('Chhattisgarh', 21107276.0, 27177482),
 ('Daman and Diu', 104550.0, 243247),
 ('Delhi', 171427.0, 142004),
 ('Goa', 1338494.0, 1458545),
 ('Gujarat', 44906170.0, 55921830),
 ('Haryana', 19942560.0, 25351462),
 ('Himachal Pradesh', 9274005.0, 10600952),
 ('Jammu and Kashmir', 9508221.0, 12541302),
 ('Jharkhand', 24691099.0, 31923078),
 ('Karnataka', 50869185.0, 61350527),
 ('Kerala', 31709259.0, 33406061),
 ('Lakshadweep', 60411.0, 64473),
 ('Madhya Pradesh', 53803203.0, 67681327),
 ('Maharashtra', 93089485.0, 113322979),
 ('Manipur', 2088188.0, 2855794),
 ('Meghalaya', 1963149.0, 2708049),
 ('Mizoram', 821407.0, 1097206),
 ('Nagaland', 63109.0, 1978502),
 ('Orissa', 36035426.0, 41974218),
 ('Puducherry', 887086.0, 1247953),
 ('Punjab', 23160779.0, 26748710),
 ('Rajasthan', 56160342.0, 71757578),
 ('Tamil Nadu', 60462105.0, 72147030),
 ('Tripura', 3125588.0, 3673917),
 ('Uttar Pradesh', 156410825.0, 196779476),
 ('Uttarakhand', 6010896.0, 7508599),
 ('West Bengal', 78426700.0, 91276115)]

##  16. Top three district of every states with highest literacy rate


In [28]:
%%sql
select a.* from
(select district,state, literacy ,rank() over(partition by state order by literacy 
desc ) as rnk from Dataset_1) as a
where a.rnk in(1,2,3) 

 * sqlite:///Indian_census_dataset.db
Done.


[('South Andaman', 'Andaman And Nicobar Islands', 89.13, 1),
 ('North And Middle Andaman', 'Andaman And Nicobar Islands', 83.91, 2),
 ('Nicobars', 'Andaman And Nicobar Islands', 78.06, 3),
 ('Hyderabad', 'Andhra Pradesh', 83.25, 1),
 ('Rangareddy', 'Andhra Pradesh', 75.87, 2),
 ('West Godavari', 'Andhra Pradesh', 74.63, 3),
 ('Papumpare', 'Arunachal Pradesh', 79.95, 1),
 ('Lower Subansiri', 'Arunachal Pradesh', 74.35, 2),
 ('East Siang', 'Arunachal Pradesh', 72.54, 3),
 ('Kamrup Metropolitan', 'Assam', 88.71, 1),
 ('Jorhat', 'Assam', 82.15, 2),
 ('Sivasagar', 'Assam', 80.41, 3),
 ('Rohtas', 'Bihar', 73.37, 1),
 ('Patna', 'Bihar', 70.68, 2),
 ('Bhojpur', 'Bihar', 70.47, 3),
 ('Chandigarh', 'Chandigarh', 86.05, 1),
 ('Durg', 'Chhattisgarh', 79.06, 1),
 ('Dhamtari', 'Chhattisgarh', 78.36, 2),
 ('Rajnandgaon', 'Chhattisgarh', 75.96, 3),
 ('Dadra and Nagar Haveli', 'Dadra and Nagar Haveli', 76.24, 1),
 ('Daman', 'Daman and Diu', 88.07, 1),
 ('Diu', 'Daman and Diu', 83.46, 2),
 ('East Delhi', 'Delhi', 89.31, 1),
 ('New Delhi', 'Delhi', 88.34, 2),
 ('South West Delhi', 'Delhi', 88.28, 3),
 ('North Goa', 'Goa', 89.57, 1),
 ('South Goa', 'Goa', 87.59, 2),
 ('Surat', 'Gujarat', 85.53, 1),
 ('Ahmadabad', 'Gujarat', 85.31, 2),
 ('Anand', 'Gujarat', 84.37, 3),
 ('Gurgaon', 'Haryana', 84.7, 1),
 ('Panchkula', 'Haryana', 81.88, 2),
 ('Ambala', 'Haryana', 81.75, 3),
 ('Hamirpur', 'Himachal Pradesh', 88.15, 1),
 ('Una', 'Himachal Pradesh', 86.53, 2),
 ('Kangra', 'Himachal Pradesh', 85.67, 3),
 ('Jammu', 'Jammu and Kashmir', 83.45, 1),
 ('Samba', 'Jammu and Kashmir', 81.41, 2),
 ('Leh', 'Jammu and Kashmir', 77.2, 3),
 ('Ranchi', 'Jharkhand', 76.06, 1),
 ('Purbi Singhbhum', 'Jharkhand', 75.49, 2),
 ('Dhanbad', 'Jharkhand', 74.52, 3),
 ('Dakshina Kannada', 'Karnataka', 88.57, 1),
 ('Bangalore', 'Karnataka', 87.67, 2),
 ('Udupi', 'Karnataka', 86.24, 3),
 ('Kottayam', 'Kerala', 97.21, 1),
 ('Pathanamthitta', 'Kerala', 96.55, 2),
 ('Ernakulam', 'Kerala', 95.89, 3),
 ('Lakshadweep', 'Lakshadweep', 91.85, 1),
 ('Jabalpur', 'Madhya Pradesh', 81.07, 1),
 ('Indore', 'Madhya Pradesh', 80.87, 2),
 ('Bhopal', 'Madhya Pradesh', 80.37, 3),
 ('Mumbai Suburban', 'Maharashtra', 89.91, 1),
 ('Mumbai City', 'Maharashtra', 89.21, 2),
 ('Nagpur', 'Maharashtra', 88.39, 3),
 ('Imphal West', 'Manipur', 86.08, 1),
 ('Churachandpur', 'Manipur', 82.78, 2),
 ('Imphal East', 'Manipur', 81.95, 3),
 ('East Khasi Hills', 'Meghalaya', 84.15, 1),
 ('West Khasi Hills', 'Meghalaya', 77.87, 2),
 ('Ri Bhoi', 'Meghalaya', 75.67, 3),
 ('Serchhip', 'Mizoram', 97.91, 1),
 ('Aizawl', 'Mizoram', 97.89, 2),
 ('Champhai', 'Mizoram', 95.91, 3),
 ('Mokokchung', 'Nagaland', 91.62, 1),
 ('Wokha', 'Nagaland', 87.69, 2),
 ('Zunheboto', 'Nagaland', 85.26, 3),
 ('Khordha', 'Orissa', 86.88, 1),
 ('Jagatsinghapur', 'Orissa', 86.59, 2),
 ('Cuttack', 'Orissa', 85.5, 3),
 ('Mahe', 'Puducherry', 97.87, 1),
 ('Karaikal', 'Puducherry', 87.05, 2),
 ('Puducherry', 'Puducherry', 85.44, 3),
 ('Hoshiarpur', 'Punjab', 84.59, 1),
 ('Mohali', 'Punjab', 83.8, 2),
 ('Jalandhar', 'Punjab', 82.48, 3),
 ('Kota', 'Rajasthan', 76.56, 1),
 ('Jaipur', 'Rajasthan', 75.51, 2),
 ('Jhunjhunun', 'Rajasthan', 74.13, 3),
 ('East Sikkim', 'Sikkim', 83.85, 1),
 ('South Sikkim', 'Sikkim', 81.42, 2),
 ('North Sikkim', 'Sikkim', 78.01, 3),
 ('Kanniyakumari', 'Tamil Nadu', 91.75, 1),
 ('Chennai', 'Tamil Nadu', 90.18, 2),
 ('Thoothukkudi', 'Tamil Nadu', 86.16, 3),
 ('West Tripura', 'Tripura', 88.69, 1),
 ('North Tripura', 'Tripura', 87.5, 2),
 ('Dhalai', 'Tripura', 85.72, 3),
 ('Gautam Buddha Nagar', 'Uttar Pradesh', 80.12, 1),
 ('Kanpur Nagar', 'Uttar Pradesh', 79.65, 2),
 ('Auraiya', 'Uttar Pradesh', 78.95, 3),
 ('Dehradun', 'Uttarakhand', 84.25, 1),
 ('Nainital', 'Uttarakhand', 83.88, 2),
 ('Chamoli', 'Uttarakhand', 82.65, 3),
 ('Purba Medinipur', 'West Bengal', 87.02, 1),
 ('Kolkata', 'West Bengal', 86.31, 2),
 ('North Twenty Four Parganas', 'West Bengal', 84.06, 3)]

# Key Findings:

### 1. Population of india - 1210854977

### 2. Average growth of india - 19.25

### 3. Average growth of different states of india :-

#####         a. Nagaland- 82.28
#####        b. Dadra and Nagar Haveli - 55.88
#####       c. Daman and Diu - 42.74

### 4. Average sex ratio of different states:-

#####         a. Kerala-	1080.0
#####        b. Puducherry-	1075.0
#####       c. Uttarakhand-	1010.0

### 5. List of the states which starting with 'a' and ending with 'm' :-

#####            a. Assam

### 6. Number of males and females in different states 

### 7. Total literacy ratio of different states

### 8. Population of different states in previous census

### 9. Top three district of every states with highest literacy rate

# Conclusion:

The analysis of Indian census data using SQL has provided valuable insights into the demographic landscape and socioeconomic indicators of the country. By leveraging powerful querying and data manipulation techniques, I was able to extract meaningful information and identify important patterns and trends.

In conclusion, the analysis of Indian census data using SQL has provided a comprehensive view of the country's demographics and socioeconomic landscape. The insights gained from this analysis can inform evidence-based decision-making, policy formulation, and resource allocation, with the ultimate goal of fostering sustainable and inclusive development across India.